V should require unsafe for type castings of zero

### Describe the bug

V allows type casting zero without requiring unsafe.



### Reproduction Steps

Example

`x := ClickFn(0)`

### Expected Behavior

Compiler should warn that unsafe is required (e.g. `unsafe { ClickFn(0) }`)

### Current Behavior

No warning is issued

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.9 1b136e2

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.9 da3112e.1b136e2
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.3.1, 24D70
|Processor           |8 cpus, 64bit, little endian, Apple M2
|Memory              |0.07GB/8GB
|                    |
|V executable        |/Users/mike/Documents/github/v/v
|V last modified time|2025-03-02 19:05:12
|                    |
|V home dir          |OK, value: /Users/mike/Documents/github/v
|VMODULES            |OK, value: /Users/mike/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/mike/Documents/github/kite
|                    |
|Git version         |git version 2.48.1
|V git status        |weekly.2025.07-129-g1b136e21 (4 commit(s) behind V master)
|.git/config present |true
|                    |
|cc version          |Apple clang version 16.0.0 (clang-1600.0.26.6)
|gcc version         |Apple clang version 16.0.0 (clang-1600.0.26.6)
|clang version       |Apple clang version 16.0.0 (clang-1600.0.26.6)
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 c8df4e27
|emcc version        |N/A
|glibc version       |N/A
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

@spytheman right now the checker doesn't do any checks for fn casts.

```
   5 | fn main() {
    6 | x := ClickFn('string')
      | ^
    7 | }
    8 |
================== C compilation error (from cc): ==============
cc: /tmp/v_501/a.01JNF2Q71EYC9FFKXGCED8WA75.tmp.c:1896:37: error: operand of type 'string' (aka 'struct string') where arithmetic or pointer type is required
cc:  1896 |         void (*x) (void) = ((main__ClickFn)(_SLIT("string")));
cc:       |                                            ^~~~~~~~~~~~~~~~~
cc: 1 error generated.
```

What should be allowed? I'm thinking only voidptr/nil

--- a/vlib/os/process_windows.c.v
+++ b/vlib/os/process_windows.c.v
@@ -14,7 +14,7 @@ type FN_NTSuspendResume = fn (voidptr) u64
 fn ntdll_fn(name &char) FN_NTSuspendResume {
 	ntdll := C.GetModuleHandleA(c'NTDLL')
 	if ntdll == 0 {
-		return FN_NTSuspendResume(0)
+		return unsafe { FN_NTSuspendResume(0) }
 	}
 	the_fn := FN_NTSuspendResume(C.GetProcAddress(ntdll, voidptr(name)))
 	return the_fn

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3592,13 +3592,20 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks',
 			node.pos)
 	} else if final_to_sym.kind == .function && final_from_sym.kind != .function {
-		if !(c.inside_unsafe || c.file.is_translated) {
+		if to_type.has_flag(.option) && node.expr !is ast.None {
+			c.error('casting number to Option function is not allowed, only compatible function or `none`',
+				node.pos)
+		} else if !(c.inside_unsafe || c.file.is_translated) {
 			if node.expr is ast.IntegerLiteral {
 				c.warn('casting number to function value should be done inside `unsafe{}` blocks',
 					node.pos)
 			} else if node.expr is ast.Nil {
 				c.warn('casting `nil` to function value should be done inside `unsafe{}` blocks',
 					node.pos)
+			} else if node.expr is ast.None {
+				if from_type.has_flag(.option) {
+					c.warn('cannot pass `none` to a non Option function type', node.pos)
+				}
 			} else if final_from_sym.kind != .voidptr {
 				c.error('invalid casting value to function', node.pos)
 			}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3591,6 +3591,18 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		|| c.file.is_translated) && !c.check_matching_function_symbols(final_from_sym, final_to_sym) {
 		c.error('casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks',
 			node.pos)
+	} else if final_to_sym.kind == .function && final_from_sym.kind != .function {
+		if !(c.inside_unsafe || c.file.is_translated) {
+			if node.expr is ast.IntegerLiteral {
+				c.warn('casting number to function value should be done inside `unsafe{}` blocks',
+					node.pos)
+			} else if node.expr is ast.Nil {
+				c.warn('casting `nil` to function value should be done inside `unsafe{}` blocks',
+					node.pos)
+			} else if final_from_sym.kind != .voidptr {
+				c.error('invalid casting value to function', node.pos)
+			}
+		}
 	}
 	if to_type.is_ptr() && to_sym.kind == .alias && from_sym.kind == .map {
 		c.error('cannot cast to alias pointer `${c.table.type_to_str(to_type)}` because `${c.table.type_to_str(from_type)}` is a value',

--- a/vlib/v/checker/tests/cast_fn_err.out
+++ b/vlib/v/checker/tests/cast_fn_err.out
@@ -1,34 +1,69 @@
-vlib/v/checker/tests/cast_fn_err.vv:16:7: warning: casting `nil` to function value should be done inside `unsafe{}` blocks
-   14 |     _ := unsafe { FnType(nil) }
-   15 |     _ := unsafe { FnType(0) }
-   16 |     _ := FnType(nil)
+vlib/v/checker/tests/cast_fn_err.vv:24:7: warning: casting `nil` to function value should be done inside `unsafe{}` blocks
+   22 |     // wrong ones
+   23 |     _ := FnType(foo)    
+   24 |     _ := FnType(nil)
       |          ~~~~~~~~~~~
-   17 |     _ := FnType(0)
-   18 |     _ := FnType('foo')
-vlib/v/checker/tests/cast_fn_err.vv:17:7: warning: casting number to function value should be done inside `unsafe{}` blocks
-   15 |     _ := unsafe { FnType(0) }
-   16 |     _ := FnType(nil)
-   17 |     _ := FnType(0)
+   25 |     _ := FnType(0)
+   26 |     _ := FnType('foo')
+vlib/v/checker/tests/cast_fn_err.vv:25:7: warning: casting number to function value should be done inside `unsafe{}` blocks
+   23 |     _ := FnType(foo)    
+   24 |     _ := FnType(nil)
+   25 |     _ := FnType(0)
       |          ~~~~~~~~~
-   18 |     _ := FnType('foo')
-   19 | }
-vlib/v/checker/tests/cast_fn_err.vv:12:7: error: casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks
-   10 | 
-   11 | fn main() {
-   12 |     _ := FnType(foo)
+   26 |     _ := FnType('foo')
+   27 |     _ := FnType(none)
+vlib/v/checker/tests/cast_fn_err.vv:23:7: error: casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks
+   21 | 
+   22 |     // wrong ones
+   23 |     _ := FnType(foo)    
       |          ~~~~~~~~~~~
-   13 |     _ := FnType(bar)
-   14 |     _ := unsafe { FnType(nil) }
-vlib/v/checker/tests/cast_fn_err.vv:16:14: error: `nil` is only allowed in `unsafe` code
-   14 |     _ := unsafe { FnType(nil) }
-   15 |     _ := unsafe { FnType(0) }
-   16 |     _ := FnType(nil)
+   24 |     _ := FnType(nil)
+   25 |     _ := FnType(0)
+vlib/v/checker/tests/cast_fn_err.vv:24:14: error: `nil` is only allowed in `unsafe` code
+   22 |     // wrong ones
+   23 |     _ := FnType(foo)    
+   24 |     _ := FnType(nil)
       |                 ~~~
-   17 |     _ := FnType(0)
-   18 |     _ := FnType('foo')
-vlib/v/checker/tests/cast_fn_err.vv:18:7: error: invalid casting value to function
-   16 |     _ := FnType(nil)
-   17 |     _ := FnType(0)
-   18 |     _ := FnType('foo')
+   25 |     _ := FnType(0)
+   26 |     _ := FnType('foo')
+vlib/v/checker/tests/cast_fn_err.vv:26:7: error: invalid casting value to function
+   24 |     _ := FnType(nil)
+   25 |     _ := FnType(0)
+   26 |     _ := FnType('foo')
       |          ~~~~~~~~~~~~~
-   19 | }
+   27 |     _ := FnType(none)
+   28 |     _ := ?FnType(0)
+vlib/v/checker/tests/cast_fn_err.vv:27:7: error: cannot cast `none` to `fn () bool`
+   25 |     _ := FnType(0)
+   26 |     _ := FnType('foo')
+   27 |     _ := FnType(none)
+      |          ~~~~~~~~~~~~
+   28 |     _ := ?FnType(0)
+   29 |     _ := ?FnType(nil)
+vlib/v/checker/tests/cast_fn_err.vv:28:8: error: casting number to Option function is not allowed, only compatible function or `none`
+   26 |     _ := FnType('foo')
+   27 |     _ := FnType(none)
+   28 |     _ := ?FnType(0)
+      |           ~~~~~~~~~
+   29 |     _ := ?FnType(nil)
+   30 |     _ := ?FnType(foo)
+vlib/v/checker/tests/cast_fn_err.vv:29:15: error: `nil` is only allowed in `unsafe` code
+   27 |     _ := FnType(none)
+   28 |     _ := ?FnType(0)
+   29 |     _ := ?FnType(nil)
+      |                  ~~~
+   30 |     _ := ?FnType(foo)
+   31 | }
+vlib/v/checker/tests/cast_fn_err.vv:29:8: error: casting number to Option function is not allowed, only compatible function or `none`
+   27 |     _ := FnType(none)
+   28 |     _ := ?FnType(0)
+   29 |     _ := ?FnType(nil)
+      |           ~~~~~~~~~~~
+   30 |     _ := ?FnType(foo)
+   31 | }
+vlib/v/checker/tests/cast_fn_err.vv:30:8: error: casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks
+   28 |     _ := ?FnType(0)
+   29 |     _ := ?FnType(nil)
+   30 |     _ := ?FnType(foo)
+      |           ~~~~~~~~~~~
+   31 | }

--- a/vlib/v/checker/tests/cast_fn_err.out
+++ b/vlib/v/checker/tests/cast_fn_err.out
@@ -1,12 +1,12 @@
 vlib/v/checker/tests/cast_fn_err.vv:24:7: warning: casting `nil` to function value should be done inside `unsafe{}` blocks
    22 |     // wrong ones
-   23 |     _ := FnType(foo)    
+   23 |     _ := FnType(foo)
    24 |     _ := FnType(nil)
       |          ~~~~~~~~~~~
    25 |     _ := FnType(0)
    26 |     _ := FnType('foo')
 vlib/v/checker/tests/cast_fn_err.vv:25:7: warning: casting number to function value should be done inside `unsafe{}` blocks
-   23 |     _ := FnType(foo)    
+   23 |     _ := FnType(foo)
    24 |     _ := FnType(nil)
    25 |     _ := FnType(0)
       |          ~~~~~~~~~
@@ -15,13 +15,13 @@ vlib/v/checker/tests/cast_fn_err.vv:25:7: warning: casting number to function va
 vlib/v/checker/tests/cast_fn_err.vv:23:7: error: casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks
    21 | 
    22 |     // wrong ones
-   23 |     _ := FnType(foo)    
+   23 |     _ := FnType(foo)
       |          ~~~~~~~~~~~
    24 |     _ := FnType(nil)
    25 |     _ := FnType(0)
 vlib/v/checker/tests/cast_fn_err.vv:24:14: error: `nil` is only allowed in `unsafe` code
    22 |     // wrong ones
-   23 |     _ := FnType(foo)    
+   23 |     _ := FnType(foo)
    24 |     _ := FnType(nil)
       |                 ~~~
    25 |     _ := FnType(0)

--- a/vlib/v/checker/tests/cast_fn_err.out
+++ b/vlib/v/checker/tests/cast_fn_err.out
@@ -1,0 +1,34 @@
+vlib/v/checker/tests/cast_fn_err.vv:16:7: warning: casting `nil` to function value should be done inside `unsafe{}` blocks
+   14 |     _ := unsafe { FnType(nil) }
+   15 |     _ := unsafe { FnType(0) }
+   16 |     _ := FnType(nil)
+      |          ~~~~~~~~~~~
+   17 |     _ := FnType(0)
+   18 |     _ := FnType('foo')
+vlib/v/checker/tests/cast_fn_err.vv:17:7: warning: casting number to function value should be done inside `unsafe{}` blocks
+   15 |     _ := unsafe { FnType(0) }
+   16 |     _ := FnType(nil)
+   17 |     _ := FnType(0)
+      |          ~~~~~~~~~
+   18 |     _ := FnType('foo')
+   19 | }
+vlib/v/checker/tests/cast_fn_err.vv:12:7: error: casting a function value from one function signature, to another function signature, should be done inside `unsafe{}` blocks
+   10 | 
+   11 | fn main() {
+   12 |     _ := FnType(foo)
+      |          ~~~~~~~~~~~
+   13 |     _ := FnType(bar)
+   14 |     _ := unsafe { FnType(nil) }
+vlib/v/checker/tests/cast_fn_err.vv:16:14: error: `nil` is only allowed in `unsafe` code
+   14 |     _ := unsafe { FnType(nil) }
+   15 |     _ := unsafe { FnType(0) }
+   16 |     _ := FnType(nil)
+      |                 ~~~
+   17 |     _ := FnType(0)
+   18 |     _ := FnType('foo')
+vlib/v/checker/tests/cast_fn_err.vv:18:7: error: invalid casting value to function
+   16 |     _ := FnType(nil)
+   17 |     _ := FnType(0)
+   18 |     _ := FnType('foo')
+      |          ~~~~~~~~~~~~~
+   19 | }

--- a/vlib/v/checker/tests/cast_fn_err.vv
+++ b/vlib/v/checker/tests/cast_fn_err.vv
@@ -1,0 +1,19 @@
+type FnType = fn () bool
+
+fn foo() int {
+	return 0
+}
+
+fn bar() bool {
+	return true
+}
+
+fn main() {
+	_ := FnType(foo)
+	_ := FnType(bar)
+	_ := unsafe { FnType(nil) }
+	_ := unsafe { FnType(0) }
+	_ := FnType(nil)
+	_ := FnType(0)
+	_ := FnType('foo')
+}

--- a/vlib/v/checker/tests/cast_fn_err.vv
+++ b/vlib/v/checker/tests/cast_fn_err.vv
@@ -9,11 +9,23 @@ fn bar() bool {
 }
 
 fn main() {
-	_ := FnType(foo)
-	_ := FnType(bar)
+	// acceptable ones
 	_ := unsafe { FnType(nil) }
 	_ := unsafe { FnType(0) }
+	_ := unsafe { FnType(foo) }
+	_ := unsafe { FnType(bar) }
+	_ := FnType(bar)
+	_ := ?FnType(none)
+	_ := ?FnType(bar)
+	_ := unsafe { ?FnType(foo) }
+
+	// wrong ones
+	_ := FnType(foo)
 	_ := FnType(nil)
 	_ := FnType(0)
 	_ := FnType('foo')
+	_ := FnType(none)
+	_ := ?FnType(0)
+	_ := ?FnType(nil)
+	_ := ?FnType(foo)
 }

--- a/vlib/v/checker/tests/mut_arg_different_muls_err.vv
+++ b/vlib/v/checker/tests/mut_arg_different_muls_err.vv
@@ -16,7 +16,7 @@ pub:
 fn window_resized(mut w &Window) {
 	window_width, window_height := 200, 100
 
-	if w.resize_fn != WindowResizeFn(0) {
+	if w.resize_fn != unsafe { WindowResizeFn(0) } {
 		println('fn present ${window_width} ${window_height}')
 		w.resize_fn(w, window_width, window_height)
 	}

